### PR TITLE
gen.go: collect canonical-data.json remotely

### PR DESCRIPTION
**Pull request by robphoenix**
see https://github.com/exercism/xgo/pull/606#discussion_r109292703

This is a very rough WIP, but I wanted to open it up to get feedback as this is new territory for me.
As it is now, the `Gen()` function will opt to retrieve the canonical data from a local source, but if this isn't available will then try to retrieve it remotely from the x-common Github repo. Is this best? Should it try remotely first?

Feel free to recommend better methods/style etc.

### CODE REVIEW

File: gen/gen.go [hunk 1]
```diff
@@ -119,23 +140,67 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
 	if err != nil {
 		return "", "local file", "" // no source control
 	}
 	if _, err = os.Stat(filepath.Join(c.Dir, jFile)); err != nil {
 		return "", "local file", "" // not in source control
 	}
+
 	// good.  return source control dir and commit.
-	return c.Dir, "exercism/x-common", string(bytes.TrimSpace(ori))
+	fmt.Printf("Fetching canonical data from a local source.\nHave you ensured it is up to date?\n")
```
Comment (robphoenix) [inline gen/gen.go]: I don't know about you, but I forget to do this all the time! A little reminder here won't go amiss. :smile: I also thought it would be good to indicate if the generator is collecting locally or remotely.

File: gen/gen.go [hunk 2]
```diff
@@ -119,23 +140,67 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
+	fmt.Println("Collecting remote data.")
+	resp, err := genClient.Get(fmt.Sprintf(canonicalDataURL, exercise))
+	if err != nil {
+		return []byte{}, "", "", err
+	}
+	if resp.Status != "200 OK" {
+		return []byte{}, "", "", fmt.Errorf("error fetching remote data: %s", resp.Status)
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (robphoenix) [inline gen/gen.go]: I'd like to change this to a `json.NewDecoder(resp.Body).Decode(&j) but didn't manage it before the end of the day.

File: gen/gen.go [hunk 3]
```diff
@@ -69,31 +84,45 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
 		return errors.New("unable to determine current path")
 	}
+
 	jFile := filepath.Join("exercises", exercise, "canonical-data.json")
 	// find and read the json source file
-	jPath, jOrigin, jCommit := getPath(jFile)
-	jSrc, err := ioutil.ReadFile(filepath.Join(jPath, jFile))
-	if err != nil {
-		return err
+	jPath, jOrigin, jCommit := getLocal(jFile)
```
Comment (robphoenix) [inline gen/gen.go]: I think the `getLocal` function could be refined a bit, moving `jFile` into it and returning the `ioutil.Readfile` in line 101. It then needs a different way to decide if there is a valid local source to use. :thinking:

File: gen/gen.go [hunk 4]
```diff
@@ -119,23 +148,67 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
 		return "", "local file", "" // not in source control
 	}
+
 	// good.  return source control dir and commit.
-	return c.Dir, "exercism/x-common", string(bytes.TrimSpace(ori))
+	fmt.Printf("Fetching canonical data from a local source.\nHave you ensured it is up to date?\n")
+	return c.Dir, "local exercism/x-common", string(bytes.TrimSpace(origin))
+}
+
+func getRemote(exercise string) (body []byte, jOrigin string, jCommit string, err error) {
```
Comment (robphoenix) [inline gen/gen.go]: I'm not sure about the error handling in this function.

**Comment by robphoenix** [comment]
This is still a bit of a mess, but getting better, I think?

File: gen/gen.go [hunk 5]
```diff
@@ -69,31 +84,45 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
-		// in the case of invalid JSON.  Use a real validator tool if you can't
-		// spot the problem right away.
-		return fmt.Errorf(`unexpected data structure: %v`, err)
-	}
-
-	// These fields are guaranteed to be in every problem
-	var commonMetadata struct {
-		Version string
+	// Decode the json
+	err = json.NewDecoder(bytes.NewReader(body)).Decode(&j)
```
Comment (robphoenix) [inline gen/gen.go]: I initially changed to using `json.NewDecoder` over `json.Unmarshal` because that was the recommended way to deal with the response from an `http.Get`. However as I'm essentially converting the `io.Reader` into a `[]byte` and then back again, as we have to decode it twice, once for the `j` struct and once for the `commonMetadata` struct, it should probably stay as it originally was, using `json.Unmarshal`, no?

Comment (petertseng) [inline gen/gen.go]: > it should probably stay as it originally was, using `json.Unmarshal`, no?

Right, doesn't seem to make much sense to go from `io.Reader` to `[]byte` back to `io.Reader`. Seems to make sense to keep it as `json.Unmarshal` or return an `io.Reader` from `getRemote` and `getLocal`. Probably good hygiene to close that `Reader` too (which implies it has to be a `ReadCloser`.

File: gen/gen.go [hunk 6]
```diff
@@ -119,23 +148,67 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
+		return []byte{}, "", "", fmt.Errorf("error fetching remote data: %s", resp.Status)
+	}
+	defer resp.Body.Close()
+	body, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return []byte{}, "", "", err
+	}
+	c, err := getRemoteCommit(exercise)
+	if err != nil {
+		return body, "github.com/exercism/x-common", "", nil
```
Comment (petertseng) [inline gen/gen.go]: should this return the error?

File: gen/gen.go [hunk 7]
```diff
@@ -69,31 +84,45 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
-	jPath, jOrigin, jCommit := getPath(jFile)
-	jSrc, err := ioutil.ReadFile(filepath.Join(jPath, jFile))
-	if err != nil {
-		return err
+	jPath, jOrigin, jCommit := getLocal(jFile)
+	var body []byte
+	fullPath := filepath.Join(jPath, jFile)
+	_, err := os.Stat(fullPath)
+	switch {
+	case err != nil:
```
Comment (petertseng) [inline gen/gen.go]: Thoughts on  `switch` vs `if err != nil`?

File: gen/gen.go (hunk 6 above)
Comment (robphoenix) [inline gen/gen.go]: If it does then, the the whole `Gen()` function will return the error, when all that's happened is the commit hasn't been fetched. So, under these circumstances, it can either generate the *cases_test.go* file without the commit, or if this is unacceptable, error out completely?

Comment (petertseng) [inline gen/gen.go]: I would suppose that depends on on whether the users (that'd be us) deem it acceptable.

For me I think I always expect to see the commit there.

But I'll probably keep using the x-common from the filesystem instead of from the web so I suppose my opinion can be overruled.

Comment (robphoenix) [inline gen/gen.go]: Ok, if the commit is always expected, then it should return the error.

> But I'll probably keep using the x-common from the filesystem instead of from the web 

Just out of interest, what are your reasons for this?

Comment (petertseng) [inline gen/gen.go]: >>  But I'll probably keep using the x-common from the filesystem instead of from the web
> Just out of interest, what are your reasons for this?

I thought about this and I think the most likely scenario to come up is if I'm working on a new generator and I want to run it very often to see whether it works. I wouldn't want to have to continually fetch the new JSON file from the network every time I try that, since it will lengthen the feedback loop and that makes development less pleasant.

Comment (robphoenix) [inline gen/gen.go]: That makes a lot of sense, it is the biggest issue with fetching from the network, especially if you have to deal with rural internet! Good reason to keep this using the local first, before resorting to the remote fetch. Thanks for clarifying this for me.

**Comment by robphoenix** [comment]
I'm on holiday currently, which is why this PR has stalled, but I hope t get it finished off next week.

**Comment by robphoenix** [comment]
updated commit message also

File: gen/gen.go [hunk 8]
```diff
@@ -119,23 +146,69 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
 	if err != nil {
 		return "", "local file", "" // no source control
 	}
 	if _, err = os.Stat(filepath.Join(c.Dir, jFile)); err != nil {
 		return "", "local file", "" // not in source control
 	}
+
 	// good.  return source control dir and commit.
-	return c.Dir, "exercism/x-common", string(bytes.TrimSpace(ori))
+	return c.Dir, "local exercism/x-common", string(bytes.TrimSpace(origin))
```
Comment (leenipper) [inline gen/gen.go]: Optional.  Wondering if the "local " prefix is necessary. It makes the generated file be different, and perhaps the only difference, if cases_test.go is already up to date with x-common. I suppose the same situation exists for the commit line even though this exercise did not get any changes in canonical-data.json.

Should there be a standard way to run the tests (say with an x-common git tree local) in order to let the cases_test.go have a consistent "// Source: " comment ?  Then, the "local " prefix could stay I suppose.

File: gen/gen.go [hunk 9]
```diff
@@ -119,23 +146,69 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
 }
 
-func getPath(jFile string) (jPath, jOrigin, jCommit string) {
+func getLocal(jFile string) (jPath, jOrigin, jCommit string) {
 	// Ideally draw from a .json which is pulled from the official x-common
 	// repository.  For development however, accept a file in current directory
 	// if there is no .json in source control.  Also allow an override in any
 	// case by environment variable.
-	if jPath = os.Getenv("EXTEST"); jPath > "" {
+	if jPath := os.Getenv("EXTEST"); jPath > "" {
```
Comment (tleen) [inline gen/gen.go]: If the caller has set EXTEST in env, but it (accidently) references a nonexistent file it will then keep rolling on to the remote. It does log which is helpful so you can see that the directory EXTEST references is no good as a data source. Should it be rolling on to the remote at that point or fail loud? I would think if EXTEST is defined than there was some intention of using it, in which case not having that file present may be worth shouting about?

File: gen/gen.go [hunk 10]
```diff
@@ -7,18 +7,20 @@ import (
...
 	"text/template"
+	"time"
 )
 
-// dirMetadata is the location of the x-common repository
-// on the filesystem.
-// We're making the assumption that the x-common repository
-// has been cloned to the same parent directory as the xgo
-// repository. E.g.
+// dirMetadata is the location of the x-common repository on the filesystem.
```
Comment (petertseng) [inline gen/gen.go]: It would be good if these were in a separate commit since they have nothing to do with collecting canonical data remotely.

File: gen/gen.go [hunk 11]
```diff
@@ -69,16 +85,27 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
+		log.Printf("[REMOTE] fetching %s test data\n", exercise)
+		jSrc, jOrigin, jCommit, err = getRemote(exercise)
+		if err != nil {
+			return err
+		}
 	}
 
 	// unmarshal the json source to a Go structure
-	if err = json.Unmarshal(jSrc, j); err != nil {
+	if err := json.Unmarshal(jSrc, j); err != nil {
```
Comment (petertseng) [inline gen/gen.go]: same with these `=` -> `:=`

File: gen/gen.go [hunk 12]
```diff
@@ -91,7 +118,7 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
 		Version string
 	}
 	if err := json.Unmarshal(jSrc, &commonMetadata); err != nil {
-		return fmt.Errorf(`Didn't contain version: %v`, err)
+		return fmt.Errorf(`didn't contain version: %v`, err)
```
Comment (petertseng) [inline gen/gen.go]: same

File: gen/gen.go (hunk 8 above)
Comment (petertseng) [inline gen/gen.go]: I'm also not sure local is the amount of detail we want to expose to potential readers of the generated files. including students. what difference does it make to them (or us) whether the file was sourced locally or remotely? what actions might one take differently as a result?

Comment (robphoenix) [inline gen/gen.go]: This is a good point, I think I was just focused on separating local and remote initially and didn't consider it beyond that. Will change.

File: gen/gen.go (hunk 9 above)
Comment (robphoenix) [inline gen/gen.go]: I hadn't considered this, good point.
As it is now, if this happened you would get output like so:
```
> EXTEST="doesnt_exist" go run exercises/leap/.meta/gen.go 
2017/04/15 07:00:36 [LOCAL] fetching leap test data
2017/04/15 07:00:36 [LOCAL] source: doesnt_exist
2017/04/15 07:00:36 [LOCAL] No test data found
2017/04/15 07:00:36 [REMOTE] fetching leap test data
2017/04/15 07:00:37 [REMOTE] source: github.com/exercism/x-common
```
which let's you know that there is no test data in the file so you can then check it and fix the env if necessary. Though maybe a quick check in the `getLocal` function, that the file exists would be helpful. I'll look into it, thanks @tleen

File: gen/gen.go (hunk 10 above)
Comment (robphoenix) [inline gen/gen.go]: :+1:

**Comment by robphoenix** [comment]
I did consider a remote flag, wasn't sure if that was overkill or not. I'll add it as a separate commit/PR if I do add it.

File: gen/gen.go (hunk 9 above)
Comment (tleen) [inline gen/gen.go]: Yeah I'm not sure it's worth adding a bunch of (if flag and file not exist) stuff when it's already in the output that we have bailed to remote.

File: gen/gen.go [hunk 13]
```diff
@@ -31,7 +34,21 @@ var dirMetadata string
 // the exercise directory. Falls back to the present working directory.
 var dirExercise string
 
-// Header tells how the test data was generated, for display in the header of cases_test.go
```
Comment (petertseng) [inline gen/gen.go]: what would you say about moving the whitespace change to the formatting change commit too?

File: gen/gen.go [hunk 14]
```diff
@@ -107,7 +135,7 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
 
 	// render the Go test cases
 	var b bytes.Buffer
-	if err = t.Execute(&b, &d); err != nil {
+	if err := t.Execute(&b, &d); err != nil {
```
Comment (petertseng) [inline gen/gen.go]: probably the `=` -> `:=` as well (into the formatting change commit)

File: gen/gen.go [hunk 15]
```diff
@@ -119,23 +147,69 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
 	// case by environment variable.
-	if jPath = os.Getenv("EXTEST"); jPath > "" {
+	if jPath := os.Getenv("EXTEST"); jPath > "" {
 		return jPath, "local file", "" // override
 	}
+
 	c := exec.Command("git", "log", "-1", "--oneline", jFile)
 	c.Dir = dirMetadata
-	ori, err := c.Output()
+	origin, err := c.Output()
```
Comment (petertseng) [inline gen/gen.go]: probably the `ori` -> `origin` too (into the formatting change commit)

File: gen/gen.go [hunk 16]
```diff
@@ -119,23 +147,69 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
+	return c.Dir, "exercism/x-common", string(bytes.TrimSpace(origin))
+}
+
+func getRemote(exercise string) (body []byte, jOrigin string, jCommit string, err error) {
+	url := fmt.Sprintf(canonicalDataURL, exercise)
+	resp, err := genClient.Get(url)
+	if err != nil {
+		return []byte{}, "", "", err
+	}
+	if resp.Status != "200 OK" {
```
Comment (petertseng) [inline gen/gen.go]: I don't particularly care which of these you do, but there was the possibility of `resp.StatusCode` and `http.StatusOK`

File: gen/gen.go (hunk 4 above)
Comment (petertseng) [inline gen/gen.go]: Was there anything specific you were not sure about? We previously talked about what the error for the commit message, and I will point out one more thing about checking the status code, but the current way is OK. Anything else we should look at?

**Comment by robphoenix** [comment]
Nope don't mind any and all suggestions. I think I've got everything, but do say if I've missed anything. And there's nothing with the `getRemote` errors that I'm concerned about now.
Thanks for the top notch reviews everyone. I'll merge this this evening if there's no objections. :tada:

File: gen/gen.go [hunk 17]
```diff
@@ -119,14 +146,15 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
...
 }
 
-func getPath(jFile string) (jPath, jOrigin, jCommit string) {
+func getLocal(jFile string) (jPath, jOrigin, jCommit string) {
 	// Ideally draw from a .json which is pulled from the official x-common
 	// repository.  For development however, accept a file in current directory
 	// if there is no .json in source control.  Also allow an override in any
 	// case by environment variable.
-	if jPath = os.Getenv("EXTEST"); jPath > "" {
+	if jPath := os.Getenv("EXTEST"); jPath > "" {
```
Comment (petertseng) [inline gen/gen.go]: Don't forget to move this one to the formatting commit

*maybe* the added newlines too

Comment (robphoenix) [inline gen/gen.go]: ok

**Comment by robphoenix** [comment]
oh, and I almost forgot, I'll add a little detail regarding this to the README.
